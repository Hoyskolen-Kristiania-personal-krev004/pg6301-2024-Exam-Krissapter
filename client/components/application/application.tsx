import React, { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import { ArticleRoutes } from "../articleRoutes";
import { ArticleContext } from "../article/articleContext";
import { LoginNavLink } from "../login/loginNavLink";
import { LoginContext } from "../login/loginContext";
import "../../App.css";

export function Application() {
  const [user, setUser] = useState("");
  const [username, setUsername] = useState("");
  const GOOGLE_CLIENT_ID =
    "1012467571987-l81oipbjpa6tg3qmkk0jge7vorfrjsau.apps.googleusercontent.com";

  async function loadUser() {
    const res = await fetch("/api/login");
    if (!res.ok) {
      throw new Error(
        "Something went wrong while fetching the user " +
          res.status +
          res.statusText,
      );
    }
    const user = await res.json();
    setUsername(user.username);
    setUser(user);
  }

  useEffect(() => {
    loadUser();
  }, []);

  async function fetchArticles() {
    const res = await fetch("/api/articles");

    return res.json();
  }
  async function onNewArticle({
    headline,
    article,
    category,
    author,
  }: {
    headline: any;
    article: any;
    category: any;
    author: any;
  }) {
    await fetch("/api/articles", {
      method: "POST",
      body: JSON.stringify({ headline, article, category, author }),
      headers: {
        "Content-Type": "application/json",
      },
    }).then(function (response) {
      if (!response.ok) {
        alert("That headline already exists");
      }
    });
  }
  return (
    <LoginContext.Provider
      value={{ username, user, loadUser, client_id: GOOGLE_CLIENT_ID }}
    >
      <ArticleContext.Provider value={{ fetchArticles, onNewArticle }}>
        <div className={"container"}>
          <header>
            <h1>News</h1>
          </header>
          <nav>
            <Link to={"/"}>Front Page</Link>
            <br />
            <Link to={"/articles"}>Articles</Link>
            <br />
            <Link to={"/articles/new"}>Add Articles</Link>
            <br />
            <LoginNavLink />
          </nav>
          <main>
            <ArticleRoutes />
          </main>
        </div>
      </ArticleContext.Provider>
    </LoginContext.Provider>
  );
}
