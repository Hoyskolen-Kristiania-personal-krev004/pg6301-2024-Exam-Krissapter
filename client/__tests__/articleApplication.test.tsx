import React from "react";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { Application } from "../components/application/application";

describe("news application", () => {
  it("renders front page", () => {
    const component = renderer.create(
      <MemoryRouter>
        <Application />
      </MemoryRouter>,
    );
    expect(component).toMatchSnapshot();
  });
});
