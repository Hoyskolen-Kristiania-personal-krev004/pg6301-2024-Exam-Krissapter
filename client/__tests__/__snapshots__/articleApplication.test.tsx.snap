// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`news application renders front page 1`] = `
<div
  className="container"
>
  <header>
    <h1>
      News
    </h1>
  </header>
  <nav>
    <a
      href="/"
      onClick={[Function]}
    >
      Front Page
    </a>
    <br />
    <a
      href="/articles"
      onClick={[Function]}
    >
      Articles
    </a>
    <br />
    <a
      href="/articles/new"
      onClick={[Function]}
    >
      Add Articles
    </a>
    <br />
    <a
      href="/login"
      onClick={[Function]}
    >
      Log In
    </a>
  </nav>
  <main>
    <div>
      Spinner
    </div>
  </main>
</div>
`;
